### PR TITLE
[improvement] Stop using lambda in doFirst to improve build caching

### PR DESCRIPTION
**Pull request by lycarter**
## Before this PR

Copying conjure sources was not cacheable because gradle does not (currently) support Java lambdas: https://github.com/gradle/gradle/issues/5510

bluf of the issue: Java lambdas do not reproducibly generate the same ID internally, so gradle is unable to determine if the lambda has changed. The easy fix is to just use an anonymous function until gradle fixes this upstream.

## After this PR

Copying conjure sources is now cacheable

relevant to https://github.com/gradle/gradle/issues/5510

### CODE REVIEW

**Comment by carterkozak** [comment]
@lycarter Any chance you can write a test that fails without this, and passes with the fix? That way we can be confident we don't reintroduce this break. We tests for similar problems, see `compileConjure does not run tasks if up to date` in `ConjurePluginTest.groovy`

**Comment by lycarter** [comment]
@cakofony this task is already in that test: https://github.com/palantir/gradle-conjure/blob/develop/gradle-conjure/src/test/groovy/com/palantir/gradle/conjure/ConjurePluginTest.groovy#L241

The test passes on Gradle 4.10.2 (what this repo is on), and I believe will fail on Gradle >= 5.0, per some of the comments on that linked gradle issue. I tried testing that out to confirm, but a large pile of other tests fail when I tried upgrading to 5.1 :/

I can dig into upgrading to upgrading to >=5.0, but it'll probably take a little while

**Comment by lycarter** [comment]
Dug in a bit further on the issues with upgrading to Gradle 5.0: they all seem to be because of methods that were deprecated in Gradle 5.0, and there's a check to ensure no deprecated methods are used. Most of these methods are pulled in from `gradle-baseline`, which is on 5.2 but doesn't have the deprecation check :(

**Comment by lycarter** [comment]
lame, a Gradle 5 upgrade for this repo (while deprecation checks are enabled) is blocked on https://github.com/palantir/gradle-baseline/pull/493 :(

**Comment by jdm2212** [comment]
@cakofony this PR can't be tested because it's fixing an issue with a future version of Gradle (and, once you upgrade to that future version, this will be tested by existing tests). In the interim, the caching problem is making `gw checkstyleMain` take ~40s on our internal project (your old project, incidentally :) we can share build scans on internal comms), so it'd be great if we could merge this and get a new release so that we can cut down our build times.

**Comment by carterkozak** [comment]
I defer to the other service infnra folks who have stronger gradle-fu than myself :-) The change looks reasonable to me though!

**Review by iamdanfox — APPROVED** [review]
:+1: thanks @lycarter, I did a similar thing recently (https://github.com/palantir/gradle-conjure/pull/108) but clearly missed this case!

**Pull request merged**
